### ISSUE TITLE
Diffractometer - Testing environment - Specifications.

### ISSUE THREAD

**Issue by paulau** [mxcube/mxcubeqt#216]
Dear all,

To understand the whole beamline, one needs to understand its parts.
First element, which I needed to understand was a motor.
Second element was a camera.

Our diffractometer is built from 6 motors and at least 1 camera. We use camera, to set
positions of all motors of the diffractometer. We can set(get) parameters of all motors also independent on camera. We get frames from a camera with some fps via a video stream.

To get software working, I need to test it. When there were no users, I could do whatever I like, but the normal operation will start soon. I need a Testing environment:

Instead of each "motor Device Server", we  have "virtual Motor Device Server",
which should simply react on commands, as normal motor would react (e.g. pass via acceleration phase, via motion with constant velocity, deceleration phase), but without any motion of any physical motor  

Instead of a real camera, I would like to have a "virtual Camera Server", which would generate a video stream with a typical picture, seen in a real beamline. The picture is simply a Sample(some loop of a thin wire attached to a metal cylinder). The "virtual Camera server" should permanently ask all "virtual motors of a diffractometer"(or may be ask real motors, if available) and set according position and orientation of a 3D model of the sample (cylinder + loop of a wire). A 2D frame - a projection  of the 3D  Object on a plane must be generated permanently e.g with fps 25 as real cameras do it.

The first Idea is to implement such a "virtual camera server"  using c++ with OpenGL (on Debian). 

Questions:
1) Do You have such virtual  visualization of Diffractometer in Testing Environment for MXCube?
2) If not, then would You like to have it in MXCube?
3) Do you have better suggestions about choice of tools: c++ OpenGL? (it seems most natural to virtualize cameras in DESY)

We came to the point during our meeting, that we "restart" from specifications of two parts of a beamline: a diffractometer and a sample changer.

May be, if we have "specifications of a diffractometer", then it could be possible to
implement "a virtual camera" for a given specifications. 

Sorry for this long text. I need just answers Yes-No on the questions 1) and  2) and recommendations, if any for question 3).

Greetings,
Pavel

**Comment by rhfogh** [comment]
Hi Pavel,

Your virtual camera server sounds great. Whether it would also work you would have to ask the people who have beamlines. 

At Global Phasing we are also working on improving the test environment. After a certain amount of fixing I can run our entire workflow in mock mode and get actual (simulated) images that can be processed afterwards.

So far I guess people have mostly used the mock objects as testing vehicles. Modulo bug fixes they seem to work fairly well, for instance operations are not immediate but take noticeable time to finish. I have done a certain amount of work with these mock objects, to make sure that if you set the position of a mock motor, get_position() gets it back again, that if you set the mock resolution the mock detector distance changes, that the mock centring procedure 1) works, 2) sets a new (if random) value  of the translation motors. Etc., all things I need to test the workflow off line. Global Phasing also has a diffraction emulator that will simulate images based on a series of test crystals, taking into account crystal orientation, diffractometer position etc., and using the standard MXCuBE machinery right until the diffraction_hook function is called. I have not tested it with standard data collection, but that could be done. And we have some long-term ideas of how you could improve the set-up, e.g. by returning centring results based on a real (but undisclosed) crystal orientation.

Might it be worth to discuss with others (who should understand better what it would take to fully simulate a beamline and how well the existing system does it) what would be a god way forward?

Rasmus

**Comment by beteva** [comment]
Dear @paulau, you are talking in terms of device servers, which is not quite the scope of MXCuBE as not everybody has the same control system. What we have are Mockup Hardware Objects, but they do not really simulate the underlying hardware in terms of timing (as I guess you need for the acceleration/deceleration/ moving...). As for the camera, I presume you are using Tango, so what might be of interest to you is the Lima Tango Simulation device server.
Last but not least - are you implementing the roadrunner? If yes, could we talk more in details together (beteva@esrf.fr), as ESRF might be interested in the implementation as well.
Cheers Antonia

**Comment by paulau** [comment]
Dear Rasmus,  Dear Antonia,
Thanks!

>> what would be a god way forward?

There are many mechanical parts in beamline, which are observed via cameras.
Also, mechanical devices are different on different beamlines.
I can implement short virtual camera for our particular diffractometer and show, how it works (also with MXCuBE).  Seeing just Values of Parameters of devices is a bit boring.
Seeing graphical dependencies and data series like position vs time is much better for quantitative representation of motion. Camera Picture is good for qualitative representation of motion.

In case of interest, everyone could adjust the code of the "virtual Camera" for own needs(cameras). If we would collect all parts into one application (or plan in advance, who makes what), with many cameras, observing mechanical parts or fluorescent screens, then it will be kind of 3D-Game "MXCuBE".

The matter of  interest is a motion, dynamics and interactive behavior and generally - simulation of cameras, which are important part of at least the software of P11 beamline in DESY.  Cameras are also built into MXCuBE version, which Jan has modified for our beamline.

The cameras are implemented not as Tango servers, but as a Web-Server with a video-stream, which can be displayed e.g with Browser, but also included into python-qt-GUI or web-GUI. We do not control cameras, (often) so that they do not belong to tango. Sometimes, we can control cameras directly (resolution, fps, contrast) via web interface or using something like http://cam1/set.php?fps=25, but not really via Tango.

Potentially, one could simulate the pictures of all cameras! Whether they display position of a sample attached to a diffractometer, or position of a hand of a sample changer.  Any mechanical device of a beamline, observed via camera can be simulated as 3D-Object, whose position. orientation etc. is controlled via software.
According 3D Objects must be just created and their orientation-position connected to according states of motors.

About lime tango server. I should read more. Thanks.

Greetings,
Pavel

**Comment by IvarsKarpics** [comment]
Hi @paulau ,

There is hardware objects called VaporyVideo.py that can be used to simulate 3D scenes and act as a mockup camera to simulate a crystal loop.

**Comment by paulau** [comment]
Thanks Ivars,  
It seems, I hear already several times the word "Lima"! 
Since the distance between our offices is around 50 meters, it could be good, if I would come to you shortly, and you would show, how this Video Testing Environment works. 
Thanks, 
Greetings, 
Pavel

**Comment by paulau** [comment]
Dear all, 

I implemented the VirtualCamera watching a Virtual 3D object as a model of a 
Sample of a diffractometer. All in c++. I build basically two main classes. 
Something like this: 

************************************************************************
class plotter3D
    constructor(width, height, pathtoresourcefiles)
    setPositionAndOrientation(x,y,z, anglex, angley, anglez) 
    getJpegFrame(*buffer)
    saveJpegFrameToFile(fname)
    destructor()

// usage of external libraries: OpenGL, jpeglib, 
// Works now under debian. OpenGL is tightly connected to X-windows. 
// but it works without any window(and does not need it), just in background. 

************************************************************************
class ObserverDiffProxy
    constructor(parametersfilename) // 
    getDiffractometerParameters(...)
    destructor()

// usage of external libraries: Tango
 
Then I created a plugin for the rather old c software mjpeg-streamer, used in our beamline to translate 
videostream via http server. There is basically the loop like this:

************************************************************************
pl3D = new plotter3D(...)
pxy = new ObserverDiffProxy(...)
while (1)
    pxy.getDiffractometerParameters(...)
    pl3D.setPositionAndOrientation(...)
    pl3D.getFrame()
    
    ... pass the frame to http ...
    
    wait(...)

delete pl3D
delete pxy

// usage of external libraries: jpeglib 

************************************************************************

Then, i just changed the hostname of CameraWebServer in configuration 
file of GUI for our beamline (to the host of just started Virtual Camera) 
and I got the following:

http://pavelpaulau.ddns.net/demoHelicalScanSimulationEnvironment.mp4

(Sorry for slow hosting on my kitchen. I do not have another public one.
If your browser does not support mp4 (Chromium on Debian can play it),
then You may need first to download the file and open via other program).

So, if this "testing environment" is interesting for MX^3 community, I 
can commit-upload it and help with the use etc. 
It is used from python GUI via 'import http'.
Definitely it can be used via web-GUI. 


Greetings, 
Pavel

**Comment by paulau** [comment]
Sorry. The indention in interfaces description above was swallowed during "paste" operation. 
Anyway, real current interfaces are not so nice. : http://pavelpaulau.ddns.net/tmp/
In case of common use, better interface must be developed and agreed. 
Greetings,
Pavel

**Comment by paulau** [comment]
Dear all,

Beamlines are very expensive, practically always busy, under reconstruction or broken. Therefore a Simulation Environment is very important. It  allows us to test Software without actual usage of a beamline.

Developers of mxcube or my colleagues in P11 in DESY use the so called "mockup hardware objects".
I prefer however, the more realistic simulation. Specifically: each device can be simulated as
virtual Device Server. In this case, the GUI works as it would work with real Hardware(Device-Servers). We use Tango control System and it is very easy to simulate any Device, that we have. It allows better understanding of functioning of corresponding devices and to test real code of GUI instead of testing some mockups.

So, I have Virtual Motors, Virtual Registers, Virtual Macroservers (controlling several registers, motors, detector etc).  Currently - for Tango. One can take ready Simulated devices, one can compose fast own devices. As I mentioned, I implemented a  Virtual Camera. It shows 3D object, whose position and orientation and illumination is defined via the states of several virtual motors.
The simulation Environment is tested with the local GUI software "crystalControl" of the beamline p11 in DESY.

Now, I started to work with mxcube and I use the same simulation environment. 
The illustration of work of the virtual camera with MXCUBE can be found here: http://pavelpaulau.ddns.net/DemoSimulationVirtualCamMxCube.mp4  (Use Chromium to see the video). I should improve yet pixels_per_um and beamposition values, to get exact centering.

About diffractometer:
Our diffractometer is very simple. It has actually only one rotation axis, 3 translations in space and 2 translations of a sample relative to the rotation axis.
So, the only simple 3click centering procedure, which uses linear solver of numpy is written for P11 diffractometer.

According to my Chefs, I had to work with more than 2 Years old branch of mxcube. So, it will take some more time, if adjustment to actual mxcube version will be necessary.

In case, if the other cameras(also based on other control Servers e.g. tina) are interesting, one could implement them.

And: It looks like, I must leave the beamline. The Chefs were expecting something more from me, what I did not gave them.  So I will apparently not continue to work for DESY and must search some other Job.  If the  freelance for mxcube would be possible, it would be interesting for me. Otherwise, I will find something else. 

Anyway
Greetings,
Pavel

**Comment by paulau** [comment]
Deal all,

Some questions about diffractometer:

In MiniDiff.py:
        qmo.CentredPosition.set_diffractometer_motor_names("phi",
                                                           "focus",
                                                           "phiz",
                                                           "phiy",
                                                           "zoom",
                                                           "sampx",
                                                           "sampy",
                                                           "kappa",
                                                           "kappa_phi")

Can I get Info about the meaning of each name?
May be it could be good to have some help- as comments in py code. 
Specifically: I see, that in my (old) version of mxcube, the "focus" motor is connected with the motor, which moves the sample along the beam! I would understand focus differently, it should be motor, which moves camera, without moving the sample. 
The names like phix - phiy force me also to get used to them. 
From one side, phi is the Greek letter, normally used to denote angles, from other side x,y - Latin letters, used normally to denote Cartesian coordinates (kind of translations). If we combine phi and x, one could expect, that this variable means rotation angle around x-axis. It is not in reality, it is apparently just shift of sample along x. Is it?
About the name kappa_phi - also not clear. We have enough Greek letters. 
About sampy and phiy: It could mean that sampy moves the sample, but phiy does not!  Is it correct?

Summary: Actually I want to know mainly whether in other beamlines "focus" = "move sample along the beam".

Greetings, 
Pavel

**Comment by rhfogh** [comment]
Hi Pavel,

I would very much favour some kind of standardisation here - see https://github.com/mxcube/ui-api/issues/14, The issue was also discussed in the 24 May 2019 developers meeting (http://mxcube.github.io/mxcube/doc/developers_meetings/index.html).

I think I can help with some context:

- The three angles ought to be called omega, kappa, and phi. But in old single-axis goniostats they apparently used 'phi' for the single, rotating axis. After people moved to kappa goniostats they did not want to rename the rotating axis from 'phi' to 'omega', so they called the new phi axis 'kappa_phi', as in 'the phi axis that sits on the kappa'. 

- There are apparently three translation motors that move the whole goniostat. They are called phix, phiy, phiz, because they move the phi (which should be called the omega) axis. One of them has the effect of moving the sample parallel to the beam, closer-to / away-from the camera, which is used for focusing on at least some beamlines, and so it is called 'focus'.

- There are two translation motors mounted on the goniostat head, just below the pin. Those are sampx and sampy. For sample centring you use those two and one of the all-goniostat motors, I believe it is phiz (?).   

- The names are defined relative to standard (?) coordinate system definitions, and anyway they are likely to differ between beamliines, at least if the omega axis direction changes from horizontal to vertical. So which 'phi' motor serves as a centring motor, and which (if any) as the 'focus' motor is not necessarily the same everywhere.  

Standardisation or renaming of these motors would be highly desirable IMHO, but there seemed to be no general agreement on that, let alone on how it should be done or whether it was worth the time it would take. If you can help generate some momentum here I, for one, would welcome it.

**Comment by paulau** [comment]
Thanks Rasmus!

>>whether it was worth the time
I heard already these words!  Software development is a beautiful game! I never regret my time if I got pleasure from it! Whether I play a game or read nice book or just do sports or also drink alcohol with friends!

>>Standardisation or renaming of these motors would be highly desirable IMHO
good(with clear meaning) names - pleasure from work with the project. bad names - less pleasure(or even worser).

>>but there seemed to be no general agreement on that
apparently because of lack of time for it(more precisely lack of pleasure from spending the time for it)! Or because may be true, the proposed new names are not
better than old names for certain diffractometers.

>> If you can help generate some momentum here I, for one, would welcome it.
Thanks.  Below are some suggestions:

>>There are two translation motors mounted on the goniostat head,
>>just below the pin. Those are
>>sampx and sampy. For sample centring you use those two
>>and one of the all-goniostat motors, I
>>believe it is phiz (?).
Yes, I use sampx, sampy, phiz for centering, and also phiy (well very simple translation of the sample orthogonal to the beam in horizontal plane, independent on centering of the above mentioned 3 motors).

The prefix "samp" is not the best solution, indicating the role of the motors! Although phiy, phix and phiz move the whole tower, it also means, that they also move the sample! i.e. for some cases, it does not matter which motor i use phiz or sampy, I have identical result. To avoid this confusion, the better names are desirable: One could chose something of:
sampx -> x_centering
sampx -> x_centStage
phix -> x_tower
phix -> x_gonio
phix ->  x.

The proposed names are not so nice that everyone likes them immediately and all agree to use them. No. It will be apparently not the case.

From other side, I would like the above names (x_gonio and x_centStage) more for our particular Goniometer in P11.

Anyway, we have many different goniometers. Goniometers are supposed first of all to move a sample! And namely - very precisely! The preciser, the better!
may be even, the better version for names is just:
phix -> x  (just omit phi, because there is no meaning in it. it could be some "angle", but it is out)
phiy -> y
phiz -> z
sampx ->  x_axis    (shift of a sample relative the rotation axis)
sampy ->  y_axis    (shift of a sample relative the rotation axis)

From the recent experience:
Our goniometer is heavy thing, may be 3kg or more.
To move this heavy thing the stepper motors are used!
They are not as precise as other motor types, they
are not so fast or so slow(sometimes the very slow motion is needed)
as the other motor types. It could come in future, that sampx and sampy motors
will be extended via "sampz" motor, which will need to move much lighter
system and hopefully be more precise and have better ranges of velocities.

For such a system the names like (x,y,z) - tower and (x_f,y_f, z_f) f="fine", may be also (x_axis, y_axis)  would be reasonable.

The usage of "_f" or "_axis" or may be "_tower" is kind of subscripts, which allow to write some equations on paper (or in pdf documentation) easier.

May be, the old names are  also fine, but for each goniometertype, it would be good to have
a) the drawing with clear notations of all motor names. or may be even:
b) Interactive 3D-model, illustrating, what each motor does! (it is may be too
far from real scientific aims).

Ok, This message is not assumed to provide any decision, but may be it could stimulate colleagues to think about alternatives and to express opinions:
this I would like, this I would not like ...

Greetings,
Pavel

**Comment by paulau** [comment]
To summarize and extend a bit the previous message (keywords: diffractometer, names, refactoring):

currently, the best names for motors of the goniometer in my opinion are:
x
y
z
x_axis or  x_offset  or  x_centering
y_axis or  y_offset  or  y_centering
omega
phi
kappa


The convention for names would be then: Call the object as short as possible
and reflect the meaning in the name. Use "subscript" = sequence of characters
after the symbol "_". E.g. "x_axis". Avoid usage of something, what has no
meaning in the name e.g."phi" has no meaning in "phix" or its meaning is not
very relevant to the task and can be completely ignored. (Or may be I did not understood it?)
Use Greek symbols (its names) for rotation,
Use Latin symbols for translations.

Additionally, may be it is reasonable to write the code short: i.e. 
not to use suffix like _hwobj e.g. in

self.phi_motor_hwobj = self.getObjectByRole('phi')

* if it is motor, then it is clearly hardware Object in the given context,
* if it is phix, then it is most likely motor,  and not something else,
* and finally, it is "x" which is most likely the object, rather than float in this context

If you like more, then capital X could be used for objects.
==================================================
May be, the use of such methods like getObjectByRole
could go into history, one could use polymorphism and use constructor to initialize
objects:
 
self.phi = HardwareObject("phi.xml")

where constructor of HardwareObjects would get the very clear string argument,
telling, the name of the file with initialization parameters for the motor?

I would  also add polymorphism, adding additional constructors, which would understand:

self.phi = HardwareObject("settings.cfg", "phi")

Additionally, we define the "role" too many times:

===================================================
1) in NanoDiff.py  in  

class NanoDiff(HardwareObject):
...
self.phi_motor_hwobj = self.getObjectByRole('phi')

2) in nano-diff.xml:

<equipment class="NanoDiff">
...
<object href="/phi" role="phi"/>

3) in phi.xml

<device class="SardanaMotor">
...
<taurusname>phi</taurusname>

========================================================================

what meaning has additional "mapping" for phi in nano-diff.xml?
I understand in phi.xml - we can change the python class there and
the string name for the central control server!

Well, may be the last thing is ok.  I just did not got, what it gives us.

It is interesting to understand the issue about scalability!
What does it mean and how we are going to scale! Example?


Greetings,
Pavel

**Comment by rhfogh** [comment]
@paulau 

- Simplification and reorganisation of the config system sounsds like a good idea to me (when is the right time is a longer discussion, as always).

- When we do it we should also look at how the directories are organised, how to switch between mock and normal, how many places we need to specify a given object.

- Martin Savko has a good set of  motor names in his latest soleil directory - I would suggest going for those:
alignmentx, alignmenty, alignmentz, centringx, centringy, omega, kappa, phi

- On the subject of variable names and constructors, you need to think in terms of multi-person, long-term maintenance, not of the quickest and most logical treatment of the current situation. For e.g. your example:

self.phi = HardwareObject("phi.xml")

1) having a  config file name as part of the standard constructor is too highty specialised and only works in one very particular context. It would mean changing the class constructors if you ever changed the config system, and requires the sytem for finding xml files to be an integral part of object construction. Much better to have a constructor to make objects from actual parameters, and having a function to get the parameters from wherever. 

2) self.phi could be: a float value for an angle; a motor object; an axis or information holding object of some kind; the local string name for the motor; the index for the motor in some local list; ...
It is important for maintenance that it is clear what the type and content of each parameter is, lest people use the parameter for different and incompatible purposes. Since code is written only once but read repeatedly, it is much better to type a bit more in order to make the meaning clearer. Hence phi_motor_hwobj. If we can agree on a more compact system that accomplishes the same thing that is of course fine, but can we really get that implemented (AND ENFORCED) across such a large and distributed developer team??

**Comment by paulau** [comment]
Thanks!

>>alignmentx, alignmenty, alignmentz, centringx, centringy, 

excellent names!!! 

>> omega, kappa, phi 

May be there could be also meaningful "one word prefixes"  for the Greek symbols?
How one could call in other words omega? mainomega? scanomega? Even if we do not find such a words, it is so. Though, may be one could "invent" such words - words with meaning, not just Greek coding of angles!

>>If we can agree on a more compact system that accomplishes the same thing that is of course fine, 
>>but can we really get that implemented (AND ENFORCED) across such a large and distributed 
>>developer team??

I would definitely support meaningful names etc. If two or more people agree to use something in common, it is already better than everyone uses his own! If some people join to conventions, it is perfect! It means - good conventions! If some people do not like conventions, - let them be so! Individuality is also good! 

>>self.phi = HardwareObject("phi.xml")
>>having a config file name as part of the standard constructor is too highty specialised and only 
>>works in one very particular context. 

If I am not wrong, it is done nearly this way in current mxcube!?
The difference is that the special method getObjectByRole(name)  is used instead of constructor, to construct the objects! Also its argument is practically the filename!?

>>It would mean changing the class constructors if you ever 
>>changed the config system, 

If the new config system appear, the old constructor does not need to be touched! It stays working for old config system!
What is needed - to write (to extend the class via) a new constructor, which would take the other set of input arguments (all possible except of already existing one string argument) and would get configuration from new system. If the new constructor needs to have still the only one string argument, then just insert into the old constructor the check of the string. If the last 3 symbols are "xml" it means we initialize via old config system. 

>>and requires the sytem for finding xml files to be an integral part of 
>>object construction. 

True! It could be just second argument of the constructor. Path. 
"the system for finding xml" is somehow also present in the current mxcube

Meanwhile, this system is called "server"!!! I found it confusing! Server is the program which works on some computer and answers on the requests from other computers!!! If we run a server only on a local computer and the only local processes ask something, it could be also server, but could be simply application, responding on events! If we have just class, which is used within one program and responds on "requests" of the other classes, then, I would not call it server! 

>>Much better to have a constructor to make objects from actual parameters, 

It is also fine! It could be just second constructor, which takes parameters as arguments!

>>and having a function to get the parameters from wherever.

For other (c++) projects, I was working in this way!  I had special "IniParameters" class which could be initialized in different ways (from single file or from command line arguments etc), and this class was passed to constructors of different objects, so that each object could get all needed parameters, (if they are present(registered, initialized) in IniParameters object). In the end, one could print statistics: all used parameters, all unused parameters, missing parameters, help messages about each parameter(provided via config.ini files), In how many different objects certain parameter was used etc... There was also check, that the parameters are not defined twice...

May be not the best possible way, because the constructors of all objects of my project had (had to have) the same argument (object of IniParameters class), but I got pleasure to work in this way. 

Additionally, one could extend the IniParameters class via constructors, initializing the object via Database or via other filetypes, or even via interactive Dialogs etc ... 

Greetings,
Pavel

**Comment by rhfogh** [comment]
@paulau 

- On HO confriguration, the "IniParameters" class too has a separation betwween parameter setting/parsing and object initialisation. So I think we agree on this point, that the parsing code should not be part of the actual class __init__. On multiple constructors I think that makes more sense in Java or C++, where you can overload constructors and where everything is in a class anyway. In Python I would vote for a separate function (or several) to prepare the input for the constructors. Since I agree that the getByRole system etc. could do with simplification I do not think we are far apart here.

- On the mockup objects we need something that will provide the closest possible simulation, it does not matter how it is done. I would point out, though, that 1) I at least need something that will run on my laptop, and I am not sure I would want to install tango or tine servers on my laptop. 2) There is also e.g. CollectMockup, and other things that cannot be handled by device servers. 3) The resources of time and people are not unlimited - which is why it has taken so long to e.g. get the various AbstractObjects incorporated - so we need to be reasonably careful not to bite off more than we can chew. 

Would it make sense to prepare a list of possible refactoring aims (even if that will surely spook Marcus ;-) ), consider urgency and resource requirements, and discuss priorities either here or at one of the developers' meetings?

**Comment by paulau** [comment]
Thanks!

About constructor vs getByRole function. Python is really different from c++. It was just point to think about.

About tango on a laptop. It works. It will also make feeling of reality! If there are problems with control server running locally, there will be also problems on a beamline. 

We have now Raspberries.  The all beamline software could be simulated on a raspberry with Tango, Tine, virtual servers, virtual cameras, MXCUBE, centering, ... etc. Just take this Raspberry, plug it into network of the beamline, configure DHCP to give the raspberry proper IP, restart all Devices, 
all cameras, well, reconfigure a bit e.g. URLs of cameras in mxcube and control a beamline! 

Tja, Probably it is too much for Raspberry or for a laptop or for developers. 

>> a list of possible refactoring aims 

>> urgency 
>> resource requirements, 
>> priorities

Very nice keywords!
From my experience, It is heavy to "refactor" existing project. 
It could be probably faster to make the new, clean concept, split the tasks between people, 
And rewrite the project from zero, keeping the well working existing project supported. 

urgency - no, :)
resource requirements - Yes  :)
priorities - Everyone has own!  :) 

If, to choose to modify mxcube, rather than to start from a new scratch, then 
I would personally see the following  "list of possible refactoring aims " 
(ordered by decreasing priority):

1) Clean Up - remove old stuff
2) Realistic Simulation - All must work on "my" laptop (or raspberry) nearly so as it will work on a beamline! 
3) Structure - organize common mxcube code, Beamline "private" code, beamline settings, external dependencies (Blis, Tango, Tine ...?)
4) Alternative parameters initialization system
5) Blis alternatives. 
6) Multiplatform usage. (Windows? Sometimes, I hate Debian, Sometimes it is beautiful)

Well it is just first version of my vision. Definitely it can be modified extended, discussed, ...
May be tomorrow, I will think differently. :)

Discussion live is always good! :) 
It could be also good to have such "a list of possible refactoring aims " in hands 
to make live discussion more organized. 

Greetings,
Pavel

**Comment by marcus-oscarsson** [comment]
Hi,

I see that the discussion is running hot here, I've been very busy so I have not had the opportunity to answer yet, my apologies !

The refactoring that we have in mind is much more drastic, some fundamental concepts will of course remain. The API of certain things will maybe even look similar others will most likely change quite a lot. But the framework, BlissFramework , with bricks and hardware objects, and the way configuration is done will most likely change.

I think we should have a similar approach for the "core refactoring" as we have for the "UI-API" which means that we gather ideas, ask our self's if these ideas are desirable and feasible to implement.

I would also like that we take a moment to look at the process that we are using. I would consider introducing tests, code standardization and documentation as a part of this refactoring task.

We have plenty of ideas but I don't think now is the right idea to present them. I believe that its important that we focus at the current task at hand which is the "UI-API", so that we can get consensus and well informed decisions regarding that task. I think that the outcome of the "UI-API"  work will influence what we do next. We are also not so far from having something that we can start to use. So even if I for one would be very eager to start this "core refactoring" my self I think its simply not the moment right now.

I think that also @MartinSavko is working on a suggestion for the diffractometer part of "UI-API", so there might be some good discussion about naming of motors and such coming out from that. So, I would also rather wait and discuss that kind of things until we have that suggestion. (@MartinSavko not in anyways trying to stress you :), we will discuss whenever the proposal is ready ...)

Cheers,
Marcus

**Comment by paulau** [comment]
Dear all,

There are keywords "diffractometer" and "specifications" in this issue.

FIRST,
one needs to define, what is "diffractometer"?
I would like to understand it as
"A device, which scope is to move and rotate(orient) a sample."

If we have it, then we immediately can suggest some "methods", which the object should have:

nClickCentering(..arguments ..)
moveSampleIntoSavePlace(...arguments...)
moveSampleBackFromSavePlace()

as was discussed above, there is some set of motors, doing these tasks.
Our colleagues in DESY develop some crazy "diffractometers" with 17
different motors!

But:
I hear from colleagues: a diffractometer should also include a beamstop! Because apparently the beamstop is very close to a sample and apparently the motors of the beamstop are connected to the same controller as the motors of the diffractometer.
Although, I would say "beamstop" is the device of different class!
Its scope is "to allign the beamstop to prevent a detector damage"

I see, that diffractometer can also "trigger"  a detector(e.g. Pillatus). It means, that
at certain angle some hardware pin of the diffractometer will get 1, and it will be set
to 0, when the other end angle is reached.

I see also, that "diffractometer" can also get(and set) the state of the fast shutter.

Should diffractometer also contain "camera" ? Or it must belong to other class, which definitely must interact with Diffractometer(if it is just restricted by scope to move and to orient)

We can make "diffractometer" bigger and bigger, interconnected with all elements of
a beamline, then we must call it not "diffractometer" anymore, but "beamline".

And now is the question:

What is a diffractometer???

SECOND,
Where is the best place for the code of a diffractometer?
On a desktop within mxcube?
Could be. However, it could be also all done in "DeviceServer".
Advantage to do it on Server side:
a) if Diffractometer is a separate device, then it would be good, if its manufacturer
also supply "drivers" or in our language "device servers".
Then, the specifications of "diffractomerer" are specifications for the server, not for the "hardwareobject"
b) If the motion of the sample should be done very synchronously with several motors simultaneously, then it is better to control this motion "locally" not via ethernet. The same is true, if the simultaneous shutter control or trigger of a detector is needed.

One more note:
There is Centering class in mxcube. Centering inherits Device. There is also NanoDiff class, which inherits HardwareObjects which inherits Device! NanoDiff has Cantering as its property. I would say, it is enough to have NanoDiff and to provide the function Center() instead of having one more class Centering!
What is the device "Centering"? Why it is done as separate Class? Why all its functionality does not belong to the class NanoDiff (our diffractometer)? What has to do the concept of Tensor for diffractometer? Is the tensor transformation is somehow used in the centering? As far as I know we do not make linear transformations of basis and we do not need to transform tensors accordingly. I would use more simple names. If we do not need tensor properties of a matrix, then we do not use this word. 

One more thing:

I learned Tango just half a year ago, and I see in the old projects some prehistorical code, repeated permanently in thread like:

self.currentAngle = self.proxyGoniometer.read_attribute("Position").value

just to have always actual value of position of some motor in the property self.currentAngle of the QThread object self. It is kind of channel of mxcube, I guess.

Now, the  modern  proxyGoniometer.Position  provides always actual value of Position without any  threads ("self")! It means, there is no need in custom channels! Everything is already done for us.

So, If, the control system would be developed for modern Tango, Then, there is no need in channels and in commands!!! This functionality is already standard functionality of PyTango library! I do not know about Tine and other systems. It would be definitely nice to provide also according functionality in python libraries for Tine, and not in mxcube project! 

In principle, in this situation, one could exclude HardwareObjects completely!
If DeviceServer is written according to specifications, then just create Proxy Object and have all Methods and Properties for your use!!!

But: the world is heterogeneous. I need sometimes "caravan adapter" to connect my device to the power socket. I do not like cables with adapters at home!!! It is always more pleasant to plug a small and compact plug into the according socket!

Summary:
There are many possibilities and questions related to the mxcube refactoring.

Greetings,
Pavel

**Comment by pkeller** [comment]
Hi Pavel,

> one needs to define, what is "diffractometer"?
> I would like to understand it as
> "A device, which scope is to move and rotate(orient) a sample."

I don't have the time for a detailed response now, but I would like to point out that the domain-specific terms used in this field already have established meanings and definitions that cannot be arbitrarily assigned or overridden by new ones without very good reasons. In particular, "diffractometer" means "a device to measure diffraction", just as "accelerometer" means "a device to measure acceleration".

The term for the device that positions the sample is "goniostat". For historical reasons "goniometer" is often used instead, but strictly speaking a goniometer is used to _measure_ angles, whereas a goniostat is used to _set_ angles ("thermometer" and "thermostat" are a good analogy).

Regards,
Peter.

**Comment by paulau** [comment]
Thanks Peter!!!

Very nice explanation! 

Amazing! 

I did not had the chance to understand this words so nicely yet!

Well, to measure Diffraction, we need also detector!
It was not listed in my post above!
I have also not seen any methods related
to detector among the methods of a relevant
(NanoDiff) hardware Objects. Sorry, I was not really very careful, working with the code.

We can not measure Diffraction without a beam!!!
The "beam" (its size, flux, photon energy manipulation) must also belong to the DiffractoMeter!

May be, what is meant under current "diffractometer" HardwareObject
should in reality be called goniostat!

Sorry, that I write here so many stupid things.

Greetings,
Pavel

**Comment by paulau** [comment]
Ooops. 

1) Motornames
I got small notice from one of my colleagues, why the names 
alignmentx, alignmenty, alignmentz are not definitely the best choice!

Because we can use the word align for everything! 

Can we say, that centering is also alignment?
One could offer  positioningx positioningy positioningz, but centering is also positioning. 
in some sence. or may be still just x, y, z. :). Anyway, I found alignmenty better than phiy.  
One could formulate the different sets of names and ask people to vote.  :) . Or keep it as it is. 
a1 a2 a3 are also good names.

2) Diffractometer - Goniostat - ... 
Just came to the head: 
goniostat = 'angle' + 'set'. This means practically to set only orientation - not a position. 
Instead of 'goniostat' or 'diffractometer', in the meaning = 'Device to set position and orientation 
of a sample', I would simply use:  SampleHolder or just Holder. 

Summary:
It is probably too much!!! 
If people are used to call apples pears, then probably they will prefer to keep it like this. 

Greetings, 
Pavel

**Comment by paulau** [comment]
Well, on the higher level (above motors etc), it is reasonable to have a list of classes, with their names 
and the core concepts of the beamline, e.g.: 

Name:            Core concept:
GonioStat       Positioning and Orientation of a sample
BeamStop      Positioning of the Beamstop
Filter              Reducing the flux of the beam via insertion of absorbing plates
SampleCam? Cameras for getting information(posit, orient.) of a sample using visible range light
Detector        Getting 2D images in X-Ray range
BeamFocus   Focusing of X-Ray beam. 
PinHoles        Shaping beam via Pinholes. 
SlitShaper?    Shaping beam via Slits
....

For example, I think, that the usage of a motor phix (setting position of a sample in the direction along the beam) as focus  does not belong to the core concepts of class goniostat. It should belong to the class SampleCam and be implemented  via other motor. 
I am not sure, how relevant is it. X-Ray beam has also its "focus" somewhere on phix axis. Changing phix to bring a sample in focus of visible range camera, we could bring a sample out of X-Ray focus. But probably it is irrelevant. Probably, after a shift along x-direction, a sample still stays within the X-Ray focus. 

Other example: Detector could contain  Pillatus and DetectorTowerMotors. Or a DetectorTower should be separate class? Could be! Also fine for me.

If the core concept of  DiffractoMeter is "To Measure Diffraction", then it is even one level higher, than the classes listed above. 

It could be good to see the list of all classes, relevant to the project and their core concepts and UML representation of the project. Such an UML representation can be then discussed, modified, improved ...
Afterwards - coded. 

Sorry, if I wrote something stupid again. 

Greetings. 
Pavel

**Comment by paulau** [comment]
Ok, I do not put text here, because it is hard to read everything after formatting of github.
I put some ideas about refactoring here: 
http://pavelpaulau.ddns.net/mxcuberefactoringsubject.txt
Greetings. 
P.

**Comment by paulau** [comment]
I think, I better remove my comments here and close this issue.
